File: client/src/main/java/help/HelpPanel.java
package help;

import java.awt.Color;
import java.awt.Cursor;
import java.awt.Point;
import java.awt.event.MouseEvent;
import java.awt.event.MouseListener;
import java.awt.event.MouseMotionAdapter;
import java.util.List;

import javax.swing.*;
import javax.swing.text.DefaultHighlighter;
import javax.swing.text.Highlighter;
import javax.swing.text.Highlighter.HighlightPainter;

import screen.HelpDialog;
import screen.ScreenCache;
import utils.SwingUtilsKt;

import static utils.InjectedThings.logger;

/**
 * Object representing a 'page' of the help dialog.
 */
public abstract class HelpPanel extends JPanel
{
	private String panelName = null;
	private String nodeName = null;

	public HelpPanel()
	{

	}

	private List<JTextPane> getTextFields() {
		return SwingUtilsKt.getAllChildComponentsForType(this, JTextPane.class);
	}

	public boolean contains(String searchStr)
	{
		String searchStrLowerCase = searchStr.toLowerCase();

		for (var fieldToCheck : getTextFields()) {
			try
			{
				int fieldLength = fieldToCheck.getDocument().getLength();
				String fieldText = fieldToCheck.getDocument().getText(0, fieldLength);
				String fieldTextLowerCase = fieldText.toLowerCase();

				if (fieldTextLowerCase.contains(searchStrLowerCase))
				{
					return true;
				}
			}
			catch (Throwable t)
			{
				logger.error("searchError", "Encountered error searching for " + searchStr, t);
			}
		}

		return false;
	}

	public void highlight(String searchStr)
	{
		int termLength = searchStr.length();

		HighlightPainter hlp = new DefaultHighlighter.DefaultHighlightPainter(Color.YELLOW);

		for (var pane : getTextFields())
		{
			Highlighter highlighter = pane.getHighlighter();
			int paneLength = pane.getDocument().getLength();

			try
			{
				highlighter.removeAllHighlights();
				String paneText = getDocumentText(pane);

				for (int i=0; i<paneLength - termLength + 1; i++)
				{
					String potentialSearchStr = paneText.substring(i, i+termLength);

					if (potentialSearchStr.equalsIgnoreCase(searchStr) && !searchStr.isEmpty())
					{
						highlighter.addHighlight(i, i+termLength, hlp);
					}
				}
			}
			catch (Throwable e)
			{
				logger.error("highlightError", "Error highlighting pane", e);
			}
		}
	}

	public void setTextFieldsEditable(boolean editable)
	{
		for (var pane: getTextFields()) {
			pane.setEditable(editable);
		}
	}

	public void addMouseListeners(final String... wordsToExclude)
	{
		for (var pane: getTextFields()) {
			pane.addMouseListener(new MouseListener()
			{
				@Override
				public void mouseClicked(MouseEvent arg0) 
				{
					String text = getDocumentText(pane);
					Point pt = arg0.getPoint();
					int pos = pane.viewToModel(pt);

					String word = getWordFromPosition(text, pos);

					if (isKeyWord(word, wordsToExclude))
					{
						navigateToPageBasedOnKeyWord(word);
					}
				}

				@Override
				public void mouseEntered(MouseEvent arg0) 
				{
					mouseHovered(pane, arg0, wordsToExclude);
				}

				@Override
				public void mouseExited(MouseEvent arg0) 
				{

				}

				@Override
				public void mousePressed(MouseEvent arg0) 
				{

				}

				@Override
				public void mouseReleased(MouseEvent arg0) 
				{

				}

			});

			pane.addMouseMotionListener(new MouseMotionAdapter() {
				@Override
				public void mouseMoved(MouseEvent arg0) 
				{
					mouseHovered(pane, arg0, wordsToExclude);
				}

			});
		}
	}

	private void mouseHovered(JTextPane pane, MouseEvent arg0, String[] wordsToExclude)
	{
		String text = getDocumentText(pane);
		Point pt = arg0.getPoint();
		int pos = pane.viewToModel(pt);

		String word = getWordFromPosition(text, pos);

		if (isKeyWord(word, wordsToExclude))
		{
			setCursor(Cursor.getPredefinedCursor(Cursor.HAND_CURSOR));
		}
		else
		{
			setCursor(Cursor.getPredefinedCursor(Cursor.DEFAULT_CURSOR));
		}
	}

	private String getWordFromPosition(String text, int position)
	{
		if (position < 1)
		{
			return "";
		}

		int length = text.length();

		String character = text.substring(position-1, position);

		if (!isLetter(character))
		{
			return "";
		}

		String word = character;

		if (position > 1)
		{
			int i = 1;
			String characterToTheLeft = text.substring(position-i-1, position-i);

			while (isLetter(characterToTheLeft))
			{
				try
				{
					word = characterToTheLeft + word;
					i++;
					characterToTheLeft = text.substring(position-i-1, position-i);
				}
				catch (IndexOutOfBoundsException e)
				{
					characterToTheLeft = "";
				}
			}
		}

		if (position < length)
		{
			int i = 1;
			String characterToTheRight = text.substring(position+i-1, position+i);

			while (isLetter(characterToTheRight))
			{
				try
				{
					word += characterToTheRight;
					i++;
					characterToTheRight = text.substring(position+i-1, position+i);
				}
				catch (IndexOutOfBoundsException e)
				{
					characterToTheRight = "";
				}
			}
		}

		return word.toLowerCase();
	}

	private boolean isLetter(String letter)
	{
		return letter.matches("^[a-zA-Z]+$");
	}

	private void navigateToPageBasedOnKeyWord(String keyWord)
	{
		logger.info("navigatedForWord", "Navigated for word " + keyWord);
		HelpDialog helpDialog = ScreenCache.getHelpDialog();

		if (keyWord.startsWith("bidd"))
		{
			if (panelName.contains("Entropy"))
			{
				helpDialog.setSelectionForWord("RulesEntropyBidding");
			}
			else if (panelName.contains("Vectropy"))
			{
				helpDialog.setSelectionForWord("RulesVectropyBidding");
			}
		}
		else if (keyWord.startsWith("chall"))
		{
			if (panelName.contains("Entropy"))
			{
				helpDialog.setSelectionForWord("RulesEntropyChallenging");
			}
			else if (panelName.contains("Vectropy"))
			{
				helpDialog.setSelectionForWord("RulesVectropyChallenging");
			}
		}
		else if (keyWord.equals("order"))
		{
			helpDialog.setSelectionForWord("FundamentalsTheDeck");
		}
		else if (keyWord.equals("perfect"))
		{
			helpDialog.setSelectionForWord("FundamentalsGlossary");
		}
	}

	private String getDocumentText(JTextPane pane)
	{
		try
		{
			int length = pane.getDocument().getLength();
			return pane.getDocument().getText(0, length);
		}
		catch (Throwable e)
		{
			logger.error("textPaneError", "Error getting text from pane", e);
			return null;
		}
	}

	private boolean isKeyWord(String word, String[] wordsToExclude)
	{
		for (String wordToExclude : wordsToExclude) {
            if (word.startsWith(wordToExclude) && !wordToExclude.isEmpty()) {
                return false;
            }
        }

		return word.equals("bidding") 
		  || word.startsWith("challeng") 
		  || word.equals("order")
		  || word.equals("perfect");
	}

	@Override
	public String toString()
	{
		return nodeName;
	}

	public String getPanelName()
	{
		return panelName;
	}
	public void setPanelName(String panelName)
	{
		this.panelName = panelName;
	}
	public void setNodeName(String nodeName)
	{
		this.nodeName = nodeName;
	}

	public void fireAppearancePreferencesChange() {

	}
	
	public void refresh()
	{
		//to be overridden by any pages that have dynamic content
	}
}

File: client/src/main/kotlin/help/HelpPanel.kt
package help

import java.awt.Color
import java.awt.Cursor
import java.awt.event.MouseAdapter
import java.awt.event.MouseEvent
import java.awt.event.MouseMotionAdapter
import java.util.*
import javax.swing.JPanel
import javax.swing.JTextPane
import javax.swing.text.DefaultHighlighter.DefaultHighlightPainter
import javax.swing.text.Highlighter
import screen.HelpDialog
import screen.ScreenCache
import util.Registry
import utils.CoreGlobals.logger
import utils.getAllChildComponentsForType

/** Object representing a 'page' of the help dialog. */
abstract class HelpPanel : JPanel() {
    abstract val nodeName: String

    protected open fun searchTermsToExclude(): List<String> {
        return emptyList()
    }

    protected fun finaliseComponents() {
        addMouseListeners(searchTermsToExclude())
        setTextFieldsReadOnly()
    }

    private fun getTextFields() = getAllChildComponentsForType(JTextPane::class.java)

    fun contains(searchStr: String): Boolean {
        val searchStrLowerCase = searchStr.lowercase(Locale.getDefault())

        for (fieldToCheck in getTextFields()) {
            val fieldLength = fieldToCheck.document.length
            val fieldText = fieldToCheck.document.getText(0, fieldLength)
            val fieldTextLowerCase = fieldText.lowercase(Locale.getDefault())

            if (fieldTextLowerCase.contains(searchStrLowerCase)) {
                return true
            }
        }

        return false
    }

    fun highlight(searchStr: String) {
        val termLength = searchStr.length

        val hlp: Highlighter.HighlightPainter = DefaultHighlightPainter(Color.YELLOW)

        for (pane in getTextFields()) {
            val highlighter = pane.highlighter
            val paneLength = pane.document.length

            try {
                highlighter.removeAllHighlights()
                val paneText = getDocumentText(pane)

                for (i in 0..paneLength - termLength) {
                    val potentialSearchStr = paneText.substring(i, i + termLength)

                    if (
                        potentialSearchStr.equals(searchStr, ignoreCase = true) &&
                            searchStr.isNotEmpty()
                    ) {
                        highlighter.addHighlight(i, i + termLength, hlp)
                    }
                }
            } catch (e: Throwable) {
                logger.error("highlightError", "Error highlighting pane", e)
            }
        }
    }

    private fun setTextFieldsReadOnly() {
        for (pane in getTextFields()) {
            pane.isEditable = false
        }
    }

    private fun addMouseListeners(wordsToExclude: List<String>) {
        for (pane in getTextFields()) {
            pane.addMouseListener(
                object : MouseAdapter() {
                    override fun mouseClicked(arg0: MouseEvent) {
                        val text = getDocumentText(pane)
                        val pt = arg0.point
                        val pos = pane.viewToModel(pt)

                        val word = getWordFromPosition(text, pos)

                        if (isKeyWord(word, wordsToExclude)) {
                            navigateToPageBasedOnKeyWord(word)
                        }
                    }

                    override fun mouseEntered(arg0: MouseEvent) {
                        mouseHovered(pane, arg0, wordsToExclude)
                    }
                }
            )

            pane.addMouseMotionListener(
                object : MouseMotionAdapter() {
                    override fun mouseMoved(arg0: MouseEvent) {
                        mouseHovered(pane, arg0, wordsToExclude)
                    }
                }
            )
        }
    }

    private fun mouseHovered(pane: JTextPane, arg0: MouseEvent, wordsToExclude: List<String>) {
        val text = getDocumentText(pane)
        val pt = arg0.point
        val pos = pane.viewToModel(pt)

        val word = getWordFromPosition(text, pos)

        cursor =
            if (isKeyWord(word, wordsToExclude)) {
                Cursor.getPredefinedCursor(Cursor.HAND_CURSOR)
            } else {
                Cursor.getPredefinedCursor(Cursor.DEFAULT_CURSOR)
            }
    }

    private fun getWordFromPosition(text: String, position: Int): String {
        if (position < 1) {
            return ""
        }

        val character = text.substring(position - 1, position)

        if (!isLetter(character)) {
            return ""
        }

        var word = character

        if (position > 1) {
            var i = 1
            var characterToTheLeft = text.substring(position - i - 1, position - i)

            while (isLetter(characterToTheLeft)) {
                try {
                    word = characterToTheLeft + word
                    i++
                    characterToTheLeft = text.substring(position - i - 1, position - i)
                } catch (e: IndexOutOfBoundsException) {
                    characterToTheLeft = ""
                }
            }
        }

        if (position < text.length) {
            var i = 1
            var characterToTheRight = text.substring(position + i - 1, position + i)

            while (isLetter(characterToTheRight)) {
                try {
                    word += characterToTheRight
                    i++
                    characterToTheRight = text.substring(position + i - 1, position + i)
                } catch (e: IndexOutOfBoundsException) {
                    characterToTheRight = ""
                }
            }
        }

        return word.lowercase(Locale.getDefault())
    }

    private fun isLetter(letter: String): Boolean {
        return letter.matches("^[a-zA-Z]+$".toRegex())
    }

    private fun navigateToPageBasedOnKeyWord(keyWord: String) {
        logger.info("navigatedForWord", "Navigated for word $keyWord")
        val helpDialog = ScreenCache.get<HelpDialog>()

        if (keyWord.startsWith("bidd")) {
            if (javaClass.simpleName.contains("Entropy")) {
                helpDialog.selectPane<RulesEntropyBidding>()
            } else if (javaClass.simpleName.contains("Vectropy")) {
                helpDialog.selectPane<RulesVectropyBidding>()
            }
        } else if (keyWord.startsWith("chall")) {
            if (javaClass.simpleName.contains("Entropy")) {
                helpDialog.selectPane<RulesEntropyChallenging>()
            } else if (javaClass.simpleName.contains("Vectropy")) {
                helpDialog.selectPane<RulesVectropyChallenging>()
            }
        } else if (keyWord == "order") {
            helpDialog.selectPane<FundamentalsTheDeck>()
        } else if (keyWord == "perfect") {
            helpDialog.selectPane<FundamentalsGlossary>()
        }
    }

    private fun getDocumentText(pane: JTextPane): String {
        val length = pane.document.length
        return pane.document.getText(0, length)
    }

    private fun isKeyWord(word: String, wordsToExclude: List<String>): Boolean {
        for (wordToExclude in wordsToExclude) {
            if (word.startsWith(wordToExclude) && wordToExclude.isNotEmpty()) {
                return false
            }
        }

        return word == "bidding" ||
            word.startsWith("challeng") ||
            word == "order" ||
            word == "perfect"
    }

    override fun toString() = nodeName

    protected fun useFourColours() =
        Registry.prefs[Registry.PREFERENCES_STRING_NUMBER_OF_COLOURS, Registry.TWO_COLOURS] ==
            Registry.FOUR_COLOURS

    open fun refresh() {
        // to be overridden by any pages that have dynamic content
    }
}
